rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp

```diff
@@ -320,6 +320,21 @@ bool Rir2Pir::compileBC(const BC& bc, Opcode* pos, Opcode* nextPos,
         // currently it does not pay off to put any deopts in there.
         if (callTargetFeedback.count(callee)) {
             auto& feedback = callTargetFeedback.at(callee);
+            // If this call was never executed. Might as well compile an
+            // unconditional deopt
+            if (!inPromise() && srcCode->funInvocationCount > 0 &&
+                feedback.taken == 0) {
+                // Hack to record the deoptimized call, to ensure we compile
```

Comment:
Are we using this hack anywhere?

Reply:
yes, without this for example binary tree which has something like:
```
tree = function(n, ...)
  if (n == 1)
    c(...)
  else
    tree(n-1, ...)
```
will optimize before the then-branch is executed. it deoptimizes on the then branch, but before it every ends up in another leaf branch, it will have optimized the function again. thus we will never record the `c` call and will always deoptimize there. the hack ensures that the `record_call` instruction is executed on deoptimization, thus on the next optimization the code will be correct.